python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,471 @@
+
+        # Launch the jobs
+        for job in ind_jobs:
+            self.runnerGo(job)
+
+    # Launch jobs stored in our queues. Method is blocked until both queue(s) and active jobs are empty
+    def waitFinish(self):
+        while len(self.active_jobs) != 0 or not self.runner_queue.empty() or not self.status_queue.empty():
+            # Using a thread lock, check and do things every iteration
+            with self.thread_lock:
+                self.handleActiveTests()
+
+            # sleep for just a tick or two
+            sleep(0.1)
+
+        ### Runner Pool
+        # Wait for the runner pool to empty before waiting on the status pool
+        # The runner pool and status pool can exhange jobs between each other but it is always the status pool
+        # that will have items in the queue last (that last job with a finished status to be printed)
+
+        # Close the runner_pool so jobs can no longer be scheduled for launching
+        self.runner_pool.close()
+
+        # Wait for the runner_pool to empty
+        self.runner_pool.join()
+
+        # Close the status_pool so jobs can no longer be scheduled for status printing
+        self.status_pool.close()
+
+        # Wait for the status_pool to empty
+        self.status_pool.join()
+
+    # Method to do things needed to actively running tests.
+    # NOTE: We are in a thread locked state, so be quick
+    def handleActiveTests(self):
+        # loop through all the active tests and do something
+        for tester_name in self.active_jobs:
+            tester = self.scheduled_name_to_object[tester_name]
+
+            # Handle long running tests
+            self.doLongRunningTests(tester)
+
+            # Handle timeouts
+            self.doTimeoutTests(tester)
+
+    def doLongRunningTests(self, tester):
+        if tester.getTestName() not in self.reported_jobs:
+            min_time = self.default_reported_time
+            if tester.specs.isValid('min_reported_time'):
+                min_time = float(tester.getMinReportTime())
+
+            if clock() - tester.getStartTime() > min_time:
+                self.reported_jobs.add(tester.getTestName())
+                tester.setStatus('RUNNING...', tester.bucket_pending)
+                self.harness.handleTestStatus(tester)
+
+    def doTimeoutTests(self, tester):
+        if tester.getTestName() not in self.reported_jobs:
+            if clock() - tester.getStartTime() > float(tester.getMaxTime()):
+                self.reported_jobs.add(tester.getTestName())
+                tester.setStatus('TIMEOUT', tester.bucket_fail)
+                self.harness.handleTestStatus(tester)
+
+    # TODO: Possibly look into semaphore functionality for slot allocation
+    def checkAvailableSlots(self, tester):
+        # See if we can fit this job into our busy schedule
+        if self.workers_in_use + tester.getProcs(self.options) <= self.job_slots:
```

Comment:
`self.workers_in_use` isn't thread safe. I can see that you are locking the call to this routine but again that's bad style. Most of what's in here can be moved to the schedule tab or be done somewhere before we get to the threading region. For instance the oversized calculation doesn't need to happen here. Try to minimize the work and variable you need to access when working with threads.